### PR TITLE
Refactored Database Preferences

### PR DESCRIPTION
**Pull request by Cyberboss**
Gonna move these over to an entry system similar to configs with support for loading from (but not saving to) savefiles. It's just a json blob for each slot.

:cl:
code: Modularized character preferences
tweak: Character slots can be created and deleted (Up to the old limits)
tweak: Game settings are now stored per character slot
config: ALLOW_METADATA has been removed
server: Preferences are now stored in the new preferences table
/:cl:

Server crashes will result in unsaved prefs. The alternative is to run the insert/update every time a user clicks "Save"

### CODE REVIEW

**Comment by Shadowlight213** [comment]
:-1: since tg is one of the few codebases thay can actually be run locally for testing stuff/just for fun without needing to spin up a database as well.

**Comment by JJRcop** [comment]
Why not queue up saves and do them every five minutes?

**Comment by DaxDupont** [comment]
This is a dumb idea, don't do it. 
It's gonna be hella annoying for everyone who runs a private server occasionally.

**Comment by nfreader** [comment]
👍 More stats for the stat gods

**Comment by MrStonedOne** [comment]
@cyberboss batch database entries using `insert into ... on duplicate key update` style queries built with the one proc in dbcore used to making batch queries from a list.

Than use unique override timers to delay commiting to database until like 60 seconds goes by without new items to be added

**Comment by MrStonedOne** [comment]
Everybody else:

Me and cyberboss are working to either have a fallback system, or maybe create a standalone database program you can just download and launch without installing that runs an already prepared database in a standalone mariadb server.

File: code/modules/client/preferences/preferences.dm [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+/datum/preferences/proc/Load()
+	if(IsGuestKey(ckey_owner))
+		slots = list(new /datum/preference_slot)
+		return
+
+	if(SSdbcore.Connect())
+		var/datum/DBQuery/Q = SSdbcore.NewQuery("SELECT id, json FROM [format_table_name("preferences")] WHERE owner = '[ckey_owner]'")
+		if(Q.Execute() && Q.NextRow())
+			do
```
Comment (Jordie0608) [inline code/modules/client/preferences/preferences.dm]: cmon

Comment (Cyberboss) [reply code/modules/client/preferences/preferences.dm]: https://github.com/tgstation/tgstation/blob/a82690cd5e075b2856ff598fe673ca6ba72c95ab/code/controllers/subsystem/dbcore.dm#L78-L79

:?

Comment (Cyberboss) [reply code/modules/client/preferences/preferences.dm]: oh whoops

**Comment by Fox-McCloud** [comment]
>Me and cyberboss are working to either have a fallback system, or maybe create a standalone database program you can just download and launch without installing that runs an already prepared database in a standalone mariadb server.

I'd definitely encourage a simple stand alone; relying on the old system as a fallback will "work", but getting away from using the built in save files and relying more on the database will be a win for everyone involved, in the long run, and it won't be that bad if it's just "launch this program with TGstation to get it to work", then a fallback system is just not needed.

File: code/modules/client/preferences/preferences.dm (hunk 1 above)
Comment (Jordie0608) [reply code/modules/client/preferences/preferences.dm]: Tbh it was more directed at the do-while loop and lame var names.

**Comment by Shadowlight213** [comment]
@Cyberboss You might want to add some way of keeping these all seperated per server still.
Otherwise, you're lowering the total amount of character slots players have down to a third, if they are all synced.
Plus, it runs into the issue of the characters on other servers being overwritten by the chars on whatever server @MrStonedOne decides to run the migration on.

**Comment by MrStonedOne** [comment]
@Shadowlight213 my solution to that was to not run the migration at all, but host a special byond server that has access to everybody's preferences on all 3 servers, and allow them to manually import.

**Comment by ACCount12** [comment]
More dependencies on the database, more functionality that used to _just work_ being replaced for no reason? Please don't.

**Comment by MrStonedOne** [comment]
@ACCount12 commenting without reading comments, making a point that was already addressed? I see water is in fact wet.

**Comment by ACCount12** [comment]
I only see a "solution" that's about bundling a test database with tgstation, and it doesn't seem like a good solution to me. Not at all. Why even change something like that if the system we already have in place just works?

It's important to keep the process of setting up a local test server easy. Download sources, build sources, run DMB. No external dependencies. This is just convenient, and it makes it easy for people to get into the development, as opposed to more obscure codebases that you have to run loops around just to make them stop crashing on server start.

I also doubt this bundled DB thing would work on Linux, and I doubt the downstreams would be happy about that, but I guess tgstation coderbus never gave a fuck about any of that.

**Comment by MrStonedOne** [comment]
Mysql will become more and more required because the benefits of doing so will never stop outweighing the costs.

We support running without mysql and will continue to do so, but features that save to mysql would obviously not be supported in that environment.

Meaning you can run without mysql, but you can't run save preferences without mysql.

**Comment by ACCount12** [comment]
>the benefits of doing so will never stop outweighing the costs

What are those?

**Comment by SpaceManiac** [comment]
Big 👎 for requiring a DB to save character settings. Private servers don't need any of the admin features the DB provides but *do* need the core feature which is character settings - not to mention local testing for which it's important to save, say, always playing as Captain. Given the lengths we go to avoid complicating the build process beyond hitting Ctrl+K in DreamMaker, it seems odd to ask any such hosts to run something else alongside DreamDaemon.

That said, I have no problems feeding preferences to the stat gods as long as the DB doesn't suddenly become required when it was optional before.

**Comment by Cyberboss** [comment]
I've already got some other WIP, ill come back to this after that

**Pull request closed without merging**